sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -143,145 +115,231 @@ private[sql] object OrcFilters extends OrcFiltersBase {
+              case (None, Some(_)) if canPartialPushDownConjuncts => rhs
+              case _ => None
+            }
+          case BuildSearchArgument(builder) =>
+            builder.startAnd()
+            updateBuilder(left, builder)
+            updateBuilder(right, builder)
+            builder.end()
+            ()
         }
 
       case Or(left, right) =>
-        // The Or predicate is convertible when both of its children can be pushed down.
-        // That is to say, if one/both of the children can be partially pushed down, the Or
-        // predicate can be partially pushed down as well.
-        //
-        // Here is an example used to explain the reason.
-        // Let's say we have
-        // (a1 AND a2) OR (b1 AND b2),
-        // a1 and b1 is convertible, while a2 and b2 is not.
-        // The predicate can be converted as
-        // (a1 OR b1) AND (a1 OR b2) AND (a2 OR b1) AND (a2 OR b2)
-        // As per the logical in And predicate, we can push down (a1 OR b1).
-        for {
-          _ <- createBuilder(dataTypeMap, left, newBuilder, canPartialPushDownConjuncts)
-          _ <- createBuilder(dataTypeMap, right, newBuilder, canPartialPushDownConjuncts)
-          lhs <- createBuilder(dataTypeMap, left,
-            builder.startOr(), canPartialPushDownConjuncts)
-          rhs <- createBuilder(dataTypeMap, right, lhs, canPartialPushDownConjuncts)
-        } yield rhs.end()
+        actionType match {
+          case t @ TrimUnconvertibleFilters(canPartialPushDownConjuncts) =>
+            // The Or predicate is convertible when both of its children can be pushed down.
+            // That is to say, if one/both of the children can be partially pushed down, the Or
+            // predicate can be partially pushed down as well.
+            //
+            // Here is an example used to explain the reason.
+            // Let's say we have
+            // (a1 AND a2) OR (b1 AND b2),
+            // a1 and b1 is convertible, while a2 and b2 is not.
+            // The predicate can be converted as
+            // (a1 OR b1) AND (a1 OR b2) AND (a2 OR b1) AND (a2 OR b2)
+            // As per the logical in And predicate, we can push down (a1 OR b1).
+            for {
+              lhs: Filter <- performAction(t, left)
+              rhs: Filter <- performAction(t, right)
+            } yield Or(lhs, rhs)
+          case BuildSearchArgument(builder) =>
+            builder.startOr()
+            updateBuilder(left, builder)
+            updateBuilder(right, builder)
+            builder.end()
+            ()
+        }
 
       case Not(child) =>
-        for {
-          _ <- createBuilder(dataTypeMap, child, newBuilder, canPartialPushDownConjuncts = false)
-          negate <- createBuilder(dataTypeMap,
-            child, builder.startNot(), canPartialPushDownConjuncts = false)
-        } yield negate.end()
+        actionType match {
+          case TrimUnconvertibleFilters(canPartialPushDownConjuncts) =>
```

Comment:
nit:
```
case t: TrimUnconvertibleFilters =>
  performAction(t.copy(canPartialPushDownConjuncts = false)).map(Not(_))
```

Reply:
Right now, it's equally valid to just create a new action since there's nothing else to copy, but I'm fine with keeping the copy for future-proofness and because it's shorter.